foxai/explainer/computer_vision/algorithm/lrp.py

```diff
@@ -149,74 +171,87 @@ def add_rules(self, model: ModelType) -> ModelType:
+        layers: List[Module],
+    ) -> torch.Tensor:
+        def modify_weights(layer: Module, inputs: Tuple[torch.Tensor, ...], _) -> None:
+            self.layer_to_initial_input[layer] = inputs[0].data
+            self.layer_to_rule[layer].modify_weights(layer, inputs[0])
+
+        try:
+            for layer in layers:
+                self.removable_handles.append(
+                    layer.register_forward_hook(modify_weights)
+                )
+
+            output = _run_forward(model, inputs, target, additional_forward_args)
+        finally:
+            self._remove_handles()
+        return output
 
 
 class LRPCVExplainer(BaseLRPCVExplainer):
     """LRP algorithm explainer."""
 
-    def create_explainer(
+    def get_relevances(
         self,
-        model: ModelType,
+        model: Module,
+        gradients: torch.Tensor,
         **kwargs,
-    ) -> Union[LRP, LayerLRP]:
-        """Create explainer object.
-
-        Args:
-            model: The forward function of the model or any
-                modification of it.
-
-        Returns:
-            Explainer object.
-        """
-        model = self.add_rules(modify_modules(model))
-
-        return LRP(model=model)
+    ) -> torch.Tensor:
+        return gradients
 
 
 class LayerLRPCVExplainer(BaseLRPCVExplainer):
     """Layer LRP algorithm explainer."""
 
-    def create_explainer(
+    def get_relevances(
         self,
-        model: ModelType,
+        model: Module,
+        gradients: torch.Tensor,
+        attribute_to_layer_input: bool = False,
         layer: Optional[LayerType] = None,
         **kwargs,
-    ) -> Union[LRP, LayerLRP]:
-        """Create explainer object.
+    ) -> torch.Tensor:
 
-        Uses parameter `layer` from `kwargs`. If not provided function will call
-        `get_last_conv_model_layer` function to obtain last `torch.nn.Conv2d` layer
-        from provided model.
-
-        Args:
-            model: The forward function of the model or any
-                modification of it.
-            layer: Layer for which attributions are computed.
-                Output size of attribute matches this layer's input or
-                output dimensions, depending on whether we attribute to
-                the inputs or outputs of the layer, corresponding to
-                attribution of each neuron in the input or output of
-                this layer.
-                Default: None
-
-        Returns:
-            Explainer object.
-
-        Raises:
-            ValueError: if model does not contain conv layers.
-        """
         if layer is None:
             layer = get_last_conv_model_layer(model=model)
 
-        model = self.add_rules(modify_modules(model))
+        if attribute_to_layer_input:
+            relevance = self.layer_to_rule[layer].relevance_input
+        else:
+            relevance = self.layer_to_rule[layer].relevance_output
 
-        return LayerLRP(model=model, layer=layer)
+        assert (
```

Comment:
I think that assert is only for debug, so perhaps this should be done using ``if`` and ``print``.

Reply:
This check is essential although I changed it to if with AttributeError, because perhaps it suits better than assertion.